### PR TITLE
lua@5.1 & lua: update luarocks to 2.4.4

### PR DESCRIPTION
**Pull request by DomT4**
- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/master/CONTRIBUTING.md)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [x] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Does your build pass `brew audit --strict <formula>` (after doing `brew install <formula>`)?

-----

I should've jumped on top of this a while ago, apologies.

### CODE REVIEW

**Comment by DomT4** [comment]
Don't merge this straight away, looking into a something.

**Comment by DomT4** [comment]
Alright, yeah, I'm going to file a patch upstream that may not be accepted but I think is reasonable enough to at least be considered.

**Comment by DomT4** [comment]
Alright, this should be good to 🚢 whenever you're happy & CI is happy @ilovezfs. In way of explanation:

* I knew upstream had fixed the old permissions problem but it wasn't in the current release as listed at the `https://keplerproject.github.io/luarocks/releases/` URL, so I updated to what that said was the latest release, which still had the permission issues.
* Checking the main `luarocks` website led me to realise the `https://keplerproject.github.io/luarocks/releases/` was no longer updating with the latest releases, and that instead releases were now being hosted at `https://luarocks.github.io/luarocks/releases/`.
* Testing that latest release confirmed the permissions fix had been folded in & is available, resolving issues such as the one `vis` upstream mentioned [in this comment](https://github.com/martanne/vis/issues/636#issuecomment-348705320).

**Comment by DomT4** [comment]
Oh, new problem. How fun. Looking into it, again.

**Comment by DomT4** [comment]
This is _"Not great"_ ™️ but my recommendation is that it is still better than the status quo, and resolves several issues not only around `luarocks` handling but with the `lua@5.3` formula, and is worth pressing ahead with.

I intend to wander upstream later today and try to initiate some kind of conversation around the issues discovered trying to get it to build the old way, and indeed test whether they're already been resolved in `master` or not.

**Comment by ilovezfs** [comment]
I think I'd prefer we not do this. Build steps in `post_install` are basically a formula bug.

**Comment by DomT4** [comment]
If necessary I’ll split this into two PRs whilst I talk to upstream, since the Lua formula here remains working as-is. It’s the versioned stuff there’s some issues around.

**Comment by DomT4** [comment]
Actually 3 PRs maybe, the Lua@5.3 patch & pkg-config updates probably shouldn’t wait much longer sigh.

**Comment by ilovezfs** [comment]
I demand at least 4 PRs.

**Comment by DomT4** [comment]
I mean, I could probably manage it if you're desperate 😉.

**Comment by ilovezfs** [comment]
Must … have … PR … It's like :potable_water:

**Comment by ilovezfs** [comment]
What is this conflicted PR, I see?

**Comment by DomT4** [comment]
Don't know to what you refer to 😉.

**Review by ilovezfs — CHANGES_REQUESTED** [review]
:wave: post_install

**Comment by ilovezfs** [comment]
So is the issue here just about the luarocks build reporting it cannot find lua?

**Comment by DomT4** [comment]
Yeah. They seem to have broken detection of versioned executables, and even if you hack around that problem you hit a second problem that it thinks the header version doesn't align with the executable version because it thinks the executable version is null, for some reason 🙃.

**Comment by ilovezfs** [comment]
I think it's just that
```
bash-3.2$ /usr/local/Cellar/lua/5.3.4/bin/lua --version
dyld: Library not loaded: /usr/local/lib/liblua.5.3.dylib
  Referenced from: /usr/local/Cellar/lua/5.3.4/bin/lua
  Reason: image not found
Trace/BPT trap: 5
bash-3.2$ exit
```
since the patch is using `HOMEBREW_PREFIX/lib/liblua.5.3.dylib` not `@LUA_PREFIX@/lib/liblua.5.3.dylib`

**Comment by DomT4** [comment]
Yeah, that was what I ran into last night poking around but ran out of time to test fixes before I had to die for the day.

**Comment by DomT4** [comment]
The `post_install` change was essentially something I knew would work and I could put together in 2 minutes, and probably wouldn't be awful until I had a moment to test things in more detail 😄.

**Comment by ilovezfs** [comment]
ah ok … I'm just saying I'm not sure they actually broke anything.

**Comment by DomT4** [comment]
Something changed upstream, I presume, as this wasn't an issue till recently, but yeah it could _just_ be amongst the detection tweaks they tripped a wire that had been sat there for us quietly for quite some time 🙈.

**Comment by ilovezfs** [comment]
The same `      s.gsub! "@LUA_PREFIX@", prefix` will probably work.

**Comment by DomT4** [comment]
I'll be back online properly in a few hours & can take a peek then, if you like.

**Comment by ilovezfs** [comment]
np

**Comment by DomT4** [comment]
If you want to leap over me by all means don't feel obligated to wait or anything, but yeah, happy to poke around later if you're not chomping at the bit to have at it.

**Comment by ilovezfs** [comment]
@DomT4 hehe conflicted again!

**Comment by DomT4** [comment]
Definitely gonna have to go have a word with whomever keeps causing conflicts with my PRs 😜.

**Comment by SteveJobzniak** [comment]
Here's a list of Lua 5.3's C-library changes:

https://www.lua.org/manual/5.3/manual.html#8

- Continuation functions now receive as parameters what they needed to get through lua_getctx, so lua_getctx has been removed. Adapt your code accordingly.
- Function lua_dump has an extra parameter, strip. Use 0 as the value of this parameter to get the old behavior.
- Functions to inject/project unsigned integers (lua_pushunsigned, lua_tounsigned, lua_tounsignedx, luaL_checkunsigned, luaL_optunsigned) were deprecated. Use their signed equivalents with a type cast.
- Macros to project non-default integer types (luaL_checkint, luaL_optint, luaL_checklong, luaL_optlong) were deprecated. Use their equivalent over lua_Integer with a type cast (or, when possible, use lua_Integer in your code).

Applications can't just "build against Lua 5.3 and run it with 5.2-compatibility". The 5.2-compat is about `.lua` scripts. Not about the C/API side.

Therefore a separate formula for "lua@5.2" would be needed for any applications built against the Lua 5.2 C API.

**Comment by ilovezfs** [comment]
```
iMac-TMP:homebrew-core joe$ grep -r luaL_checkint /usr/local/Cellar/lua/5.3.4_2/include
/usr/local/Cellar/lua/5.3.4_2/include/lauxlib.h:LUALIB_API lua_Integer (luaL_checkinteger) (lua_State *L, int arg);
/usr/local/Cellar/lua/5.3.4_2/include/lauxlib.h:#define luaL_checkunsigned(L,a)	((lua_Unsigned)luaL_checkinteger(L,a))
/usr/local/Cellar/lua/5.3.4_2/include/lauxlib.h:#define luaL_checkint(L,n)	((int)luaL_checkinteger(L, (n)))
/usr/local/Cellar/lua/5.3.4_2/include/lauxlib.h:#define luaL_checklong(L,n)	((long)luaL_checkinteger(L, (n)))
/usr/local/Cellar/lua/5.3.4_2/include/lua5.3/lauxlib.h:LUALIB_API lua_Integer (luaL_checkinteger) (lua_State *L, int arg);
/usr/local/Cellar/lua/5.3.4_2/include/lua5.3/lauxlib.h:#define luaL_checkunsigned(L,a)	((lua_Unsigned)luaL_checkinteger(L,a))
/usr/local/Cellar/lua/5.3.4_2/include/lua5.3/lauxlib.h:#define luaL_checkint(L,n)	((int)luaL_checkinteger(L, (n)))
/usr/local/Cellar/lua/5.3.4_2/include/lua5.3/lauxlib.h:#define luaL_checklong(L,n)	((long)luaL_checkinteger(L, (n)))
/usr/local/Cellar/lua/5.3.4_2/include/lua5.3/luaconf.h:** luaL_checkint, luaL_checklong, etc.)
/usr/local/Cellar/lua/5.3.4_2/include/luaconf.h:** luaL_checkint, luaL_checklong, etc.)
iMac-TMP:homebrew-core joe$ grep -r lua_pushunsigned /usr/local/Cellar/lua/5.3.4_2/include
/usr/local/Cellar/lua/5.3.4_2/include/lua.h:#define lua_pushunsigned(L,n)	lua_pushinteger(L, (lua_Integer)(n))
/usr/local/Cellar/lua/5.3.4_2/include/lua5.3/lua.h:#define lua_pushunsigned(L,n)	lua_pushinteger(L, (lua_Integer)(n))
/usr/local/Cellar/lua/5.3.4_2/include/lua5.3/luaconf.h:** manipulating other integer types (lua_pushunsigned, lua_tounsigned,
/usr/local/Cellar/lua/5.3.4_2/include/luaconf.h:** manipulating other integer types (lua_pushunsigned, lua_tounsigned,
iMac-TMP:homebrew-core joe$
```

**Comment by DomT4** [comment]
> The same s.gsub! "@LUA_PREFIX@", prefix will probably work.

It doesn't, FWIW. I'm still poking around locally, but sadly that's not the fix here.

**Comment by SteveJobzniak** [comment]
?

**Comment by ilovezfs** [comment]
Stale bot, please go die.

**Comment by DomT4** [comment]
If you've got ideas on resolving this without leaning on `postinstall` to do a stupid dance Stalebot, I'm all ears.

**Comment by DomT4** [comment]
So. This PR should return green now. Finally.

File: Formula/lua@5.1.rb [hunk 1]
```diff
@@ -40,14 +40,17 @@ class LuaAT51 < Formula
     end
   end
 
+  # Don't use the https://luarocks.org/releases/luarocks-x.y.z.tar.gz URL
```
Comment (ilovezfs) [inline Formula/lua@5.1.rb]: comment looks wrong

**Comment by DomT4** [comment]
The amount of wrong directions the `configure` failures sent me down was borderline ridiculous on this PR. What a grim use of time.

File: Formula/lua@5.1.rb (hunk 1 above)
Comment (DomT4) [reply Formula/lua@5.1.rb]: Oh, maybe they've fixed the redirect now. I'll check.

Comment (DomT4) [reply Formula/lua@5.1.rb]: Still there:
```
Resource: luarocks
==> Downloading https://luarocks.org/releases/luarocks-2.4.4.tar.gz
==> Downloading from http://luarocks.github.io/luarocks/releases/luarocks-2.4.4.tar.gz
/usr/bin/curl --show-error --user-agent Homebrew/1.5.12-9-g49927ed (Macintosh; Intel Mac OS X 10.13.4) curl/7.54.0 --fail --location --remote-time --continue-at - --output /usr/local/var/homebrewcache/lua--luarocks-2.4.4.tar.gz.incomplete http://luarocks.github.io/luarocks/releases/luarocks-2.4.4.tar.gz
```

Comment (ilovezfs) [reply Formula/lua@5.1.rb]: have you reported it?

Comment (DomT4) [reply Formula/lua@5.1.rb]: I can go one better and file a PR, tbh. It's a one-liner.

https://github.com/luarocks/luarocks-site/blob/b8ade0ddf390eb40631f91d57ecf4f35595858d9/nginx.conf#L102

**Comment by ilovezfs** [comment]
What was the problem?

**Comment by DomT4** [comment]
> What was the problem?

A variation of:

> The same s.gsub! "@LUA_PREFIX@", prefix will probably work.`

was indeed required as you suggested months ago, but there was also an issue where upstream apparently tweaked the header detection logic in a way that doesn't work with Homebrew's directory structure here, and that header silliness lead me down a ridiculous rabbit hole. In combination they work fine.

**Comment by DomT4** [comment]
It became obvious enough when I dropped into the debug shell for extended periods and just ran build after build after build with slight tweaks each time, but outside the debug shell a lot of the helpful messages are hidden by `brew` and you just get that simple failure message. Shouldn't have taken me this long to fix it really, but time away leads to fresh thinking or some garbage like that.

File: Formula/lua@5.1.rb [hunk 2]
```diff
@@ -40,14 +40,17 @@ class LuaAT51 < Formula
...
-    url "https://keplerproject.github.io/luarocks/releases/luarocks-2.3.0.tar.gz"
-    sha256 "68e38feeb66052e29ad1935a71b875194ed8b9c67c2223af5f4d4e3e2464ed97"
+    url "https://luarocks.github.io/luarocks/releases/luarocks-2.4.4.tar.gz"
+    sha256 "3938df33de33752ff2c526e604410af3dceb4b7ff06a770bc4a240de80a1f934"
   end
 
   def install
     # Use our CC/CFLAGS to compile.
     inreplace "src/Makefile" do |s|
+      s.gsub! HOMEBREW_PREFIX, prefix
```
Comment (ilovezfs) [inline Formula/lua@5.1.rb]: Does this actually work for a non-standard prefix?

Comment (DomT4) [reply Formula/lua@5.1.rb]: Should do, theoretically. It is just a slight variation of what we've been doing in the `lua` formula for ages, without anyone complaining about it.

File: Formula/lua@5.1.rb (hunk 1 above)
Comment (DomT4) [reply Formula/lua@5.1.rb]: https://github.com/luarocks/luarocks-site/pull/131

File: Formula/lua@5.1.rb (hunk 2 above)
Comment (ilovezfs) [reply Formula/lua@5.1.rb]: How does the replacement not fail if the prefix isn't in the file?

Comment (DomT4) [reply Formula/lua@5.1.rb]: Why wouldn't the prefix be in the file? We're the ones adding that line in the inline patch. Maybe I'm misunderstanding you? I've slept for about 4 hours over the last day 🙃.

Comment (ilovezfs) [reply Formula/lua@5.1.rb]: Ah, OK. That's what I get for looking at this on my iPhone.

Comment (DomT4) [reply Formula/lua@5.1.rb]:
```
+	$(CC) -dynamiclib -install_name HOMEBREW_PREFIX/lib/liblua.5.1.dylib \
+		-compatibility_version 5.1 -current_version 5.1.5 \
+		-o liblua.5.1.5.dylib $^
```

compared to:
```
+	$(CC) -dynamiclib -install_name @LUA_PREFIX@/lib/liblua.5.3.dylib \
+		-compatibility_version 5.3 -current_version 5.3.4 \
+		-o liblua.5.3.4.dylib $^
```
in the `lua` formula. Same principle, just a slight divergence at some point for some reason. Possibly my fault given how much I've touched these formulae over the years.

**Comment by DomT4** [comment]
Would ya look at that. A green CI build. Thought I'd never see the day here at times 🤦‍♂️.

File: Formula/lua.rb [hunk 3]
```diff
@@ -73,10 +73,10 @@ def install
 
         # This block ensures luarock exec scripts don't break across updates.
         inreplace libexec/"share/lua/5.3/luarocks/site_config.lua" do |s|
-          s.gsub! libexec.to_s, opt_libexec
-          s.gsub! include.to_s, "#{HOMEBREW_PREFIX}/include"
-          s.gsub! lib.to_s, "#{HOMEBREW_PREFIX}/lib"
-          s.gsub! bin.to_s, "#{HOMEBREW_PREFIX}/bin"
+          s.gsub! libexec, opt_libexec
+          s.gsub! include, "#{HOMEBREW_PREFIX}/include"
```
Comment (ilovezfs) [inline Formula/lua.rb]: Maybe HOMEBREW_PREFIX/"include"

**Comment by DomT4** [comment]
Upstream merged the PR so the redirect is HTTPS to HTTPS now, which is fine, so I'll push new commits to use the official URL and after that I think we're alright here if you're happy.

```
Resource: luarocks
==> Downloading https://luarocks.org/releases/luarocks-2.4.4.tar.gz
==> Downloading from https://luarocks.github.io/luarocks/releases/luarocks-2.4.4.tar.gz
/usr/bin/curl --show-error --user-agent Homebrew/1.5.12-9-g49927ed (Macintosh; Intel Mac OS X 10.13.4) curl/7.54.0 --fail --location --remote-time --continue-at - --output /usr/local/var/homebrewcache/lua--luarocks-2.4.4.tar.gz.incomplete https://luarocks.github.io/luarocks/releases/luarocks-2.4.4.tar.gz
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100  137k  100  137k    0     0  79313      0  0:00:01  0:00:01 --:--:-- 79306
Downloaded to: /usr/local/var/homebrewcache/lua--luarocks-2.4.4.tar.gz
SHA256: 3938df33de33752ff2c526e604410af3dceb4b7ff06a770bc4a240de80a1f934
```

**Review by ilovezfs — APPROVED** [review]

**Comment by ilovezfs** [comment]
Shipped! Thanks @DomT4

**Comment by DomT4** [comment]
Thanks Joe. Nice to be able to finally delete this branch, heh.

**Comment by ilovezfs** [comment]
Or just hoard branches as I do!

**Comment by DomT4** [comment]
That would be the death of what’s left of my sanity.

File: Formula/lua@5.1.rb (hunk 2 above)
Comment (xu-cheng) [reply Formula/lua@5.1.rb]: This does not look right. If I am not mistaking, this should be
```ruby
s.gsub! "HOMEBREW_PREFIX", prefix
```

Comment (DomT4) [reply Formula/lua@5.1.rb]: Don't think so? `brew unpack --patch lua@5.1` shows the blob as I expected it to be:

```
$(LUA_A): $(CORE_O) $(LIB_O)
	$(CC) -dynamiclib -install_name /usr/local/lib/liblua.5.1.dylib \
		-compatibility_version 5.1 -current_version 5.1.5 \
		-o liblua.5.1.5.dylib $^
```

Comment (xu-cheng) [reply Formula/lua@5.1.rb]: I get this for non-standard homebrew prefix.
```
brew upgrade lua@5.1
==> Upgrading 1 outdated package, with result:
lua@5.1 5.1.5_6
==> Upgrading lua@5.1 
==> Downloading https://www.lua.org/ftp/lua-5.1.5.tar.gz
######################################################################## 100.0%
Error: inreplace failed
src/Makefile:
  expected replacement of #<Pathname:/home/comp/chengxu/usr> with #<Pathname:/home/comp/chengxu/usr/Cellar/lua@5.1/5.1.5_6>

Comment (DomT4) [reply Formula/lua@5.1.rb]: I'm pretty sure I tried a string replacement in the default prefix and hit issues with it, which is why I didn't go down that route. Maybe we should just save ourselves some hassle & change it to `@LUA_PREFIX@` to match the `lua` formula.

**Pull request closed without merging**